src/Stratis.Bitcoin/Connection/PeerBanning.cs

```diff
@@ -110,19 +110,23 @@ public void BanAndDisconnectPeer(IPEndPoint endpoint, int banTimeSeconds, string
             if (peerAddresses.Count == 0)
             {
                 this.peerAddressManager.AddPeer(endpoint, IPAddress.Loopback);
-                peerAddresses.Add(this.peerAddressManager.FindPeer(endpoint));
+                PeerAddress address = this.peerAddressManager.FindPeer(endpoint);
 
-                this.logger.LogTrace("{0} added to the address manager.");
+                if (address != null)
```

Comment:
Maybe a comment about why this could be null. It seems weird that we add it to the peer address manager and it comes back null. Seems to be something to do with same source?

Reply:
What I think is actually happening here is that `this.peerAddressManager.AddPeer(endpoint, IPAddress.Loopback);` is NOT actually adding the peer to the address manager and then when we call `FindPeer` it returns a null.

We need to confirm that the address was added before we add it to the list of addresses to ban.

Reply:
Because of the method `IsRoutable` for example? agree.
Also this I  peer we want to ban we should still add it to the `peerAddresses` no?

Reply:
There can also be a race condition right? we add a peer and it gets disconnected (the fix to return the address in the add method will fix this race I presume)

Reply:
If the peer was not added to the address manager due it being routable we can't ban it as it is not in the address book ;)

Reply:
Right, because of IsRoutable or because it could somehow be one of the last nodes added for a particular source (race condition here?)

But if we want to ban it, then surely it should be added to the address book? I don't know a lot about this area, just being careful.
- What happens in a private internal network? Do we not ban nodes? 
- Could I set up a bunch of nodes such that I connect last through a peer to avoid permanent banning?

Reply:
@codingupastorm you do raise valid points. However, we dont currently have a mechanism to keep track of banned peers that are not in the address manager. I don't think it will be a big job to refactor or implement this but perhaps its outside of the scope of this PR.

@dangershony is you are happy with the PR lets merge and raise another issue to track banned peers  that are not in the address manager?

Reply:
>Could I set up a bunch of nodes such that I connect last through a peer to avoid permanent banning?
I think whitelisteing will not ban a peer